Deprecate NET_SOCKETS_POSIX_NAMES option

jukkar opened the pull request:
The `CONFIG_NET_SOCKETS_POSIX_NAMES` option is marked as deprecated in favor of using normal POSIX socket API includes found under the `include/zephyr/posix` directory. If you want to use BSD socket API calls, you need to select `POSIX_API` and use the socket headers found in the POSIX subsystem. If you do not want to or cannot enable POSIX_API, then you must use `zsock_` prefix when working with BSD socket calls.

This is initial work in order to simplify the network and Posix related config options. The use of `CONFIG_NET_SOCKETS_POSIX_NAMES` is currently confusing as it conflicts with `CONFIG_POSIX_API`.

jukkar committed:
net: sockets: Deprecate CONFIG_NET_SOCKETS_POSIX_NAMES

The CONFIG_NET_SOCKETS_POSIX_NAMES option is marked as deprecated in
favor of using normal POSIX socket API includes found under the
include/zephyr/posix directory. If you want to use BSD socket API calls,
you need to select POSIX_API and use the socket headers found in the
POSIX subsystem. If you do not want to or cannot enable POSIX_API,
then you must use zsock_ prefix when working with BSD socket calls.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: socket: Rework the zephyr/net/socket.h header

Avoid circular inclusion of socket related header files.
This way applications can still continue to include
zephyr/net/socket.h and not need to separately include
POSIX socket headers.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: tftp: Use zsock_ API functions

The library should be using internal socket API functions
so that we do not need to depend on POSIX_API inside the
network stack.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: socks: Remove socket selection

Currently the socks library does not use sockets so there
is no need to select socket symbols.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
mgmt: updatehub: Use zsock_ API functions

The library should be using internal socket API functions
so that we do not need to depend on POSIX_API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: sockets: Fix the documentation to use POSIX_API

Fix the documentation and promote POSIX_API instead of
NET_SOCKET_POSIX_NAMES

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: socket: Create zsock_fcntl and fsock_ioctl functions

Have special wrappers for zsock_fcntl and zsock_ioctl functions
so that gcc warning can be avoided.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: socket: tls_ext: Use POSIX_API

The test expects to use various POSIX APIs so enable the config.
Enabling POSIX_API requires also increase of max file descriptors.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: sockets: Remove use of NET_SOCKETS_POSIX_NAMES

The setting is deprecated so change the code to either use the
native zsock_* API or enable POSIX_API to use the BSD socket API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: Change to use CONFIG_POSIX_API

Change the sample applications that use network socket API to
use the POSIX_API config because the NET_SOCKETS_POSIX_NAMES is
deprecated. Convert also the zsock_ API calls to plain BSD
socket API calls when applicable.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: websocket: Use zsock_ API calls

Use zsock_ API calls so that we do not need to enable POSIX_API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
mgmt: mcumgr: transport: udp: Select POSIX_API

The NET_SOCKET_POSIX_NAMES is deprecated so enable POSIX_API
to use the socket API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
doc: net: Remove the mention of NET_SOCKETS_POSIX_NAMES

The option is deprecated so remove it from documentation.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on include/zephyr/net/socket.h:
```diff
@@ -890,38 +921,6 @@ static inline ssize_t recv(int sock, void *buf, size_t max_len, int flags)
-}
-
 /** POSIX wrapper for @ref zsock_sendto */
 static inline ssize_t sendto(int sock, const void *buf, size_t len, int flags,
```
I wonder, could we remove those duplicate wrappers already, and just use the POSIX headers below in case `CONFIG_NET_SOCKETS_POSIX_NAMES` is enabled as well?

jukkar replied:
If one sets `CONFIG_NET_SOCKETS_POSIX_NAMES=y` and `CONFIG_POSIX_API=n`, then there is a compilation error if these symbols are not in `socket.h`. It looks like we are stuck with these until the `NET_SOCKETS_POSIX_NAMES` is removed from the Kconfig file, and then we can remove the settings.

rlubos reviewed (commented):
I really really like the idea to finally clean this up.

jukkar committed:
doc: migration-guide-3.7: Add NET_SOCKETS_POSIX_NAMES info

Add CONFIG_NET_SOCKETS_POSIX_NAMES deprecation information.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
shell: mqtt: Avoid using POSIX function names

This commit fixes this error seen in CI so that things
work even if CONFIG_POSIX_API is enabled.

subsys/shell/backends/shell_mqtt.c:727:12: error:
  conflicting types for 'write'; have
 'int(const struct shell_transport *, const void *, size_t,  size_t *)'
  727 | static int write(const struct shell_transport *transport,
                         const void *data, size_t length)

include/zephyr/posix/unistd.h:230:9: note: previous declaration
 of 'write' with type
 'ssize_t(int,  const void *, size_t)'
  230 | ssize_t write(int file, const void *buffer, size_t count);

subsys/shell/backends/shell_mqtt.c:787:12: error:
 conflicting types for 'read'; have
 'int(const struct shell_transport *, void *, size_t,  size_t *)'
  787 | static int read(const struct shell_transport *transport,
                        void *data, size_t length, size_t *cnt)

include/zephyr/posix/unistd.h:231:9: note: previous declaration
 of 'read' with type
 'ssize_t(int,  void *, size_t)'
  231 | ssize_t read(int file, void *buffer, size_t count);

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: echo-server: Adjust CONFIG_MAX_THREAD_BYTES

This fixes this error printed by CI:

Too many thread objects (21)
Increase CONFIG_MAX_THREAD_BYTES to 3

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: lwm2m: Fix zsock_fcntl impl name

The syscall name has _impl postfix so adjusting the stub
name.

zephyr-sdk-0.16.5-1/x86_64-zephyr-elf/bin/../lib/gcc/
  x86_64-zephyr-elf/12.2.0/../../../../x86_64-zephyr-elf/bin/ld.bfd:
  app/libapp.a(lwm2m_engine.c.obj): in function `zsock_fcntl_impl':
.../syscalls/socket.h:363: undefined reference to `z_impl_zsock_fcntl_impl'

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: sockets: can: Fix the test termination criteria

If enabling CONFIG_POSIX_API, we get three file descriptors open
before any socket descriptors. The sample exit criteria checks
socket 0 and 1 which is no longer a valid one, so change the
check to be either 0 or 3, and 1 or 4 so that the test can be
run in both when POSIX API is enabled or disabled.

Also disable color printing for the test run so that it is
easier to read the output log.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: sockets: Use only Zephyr POSIX headers

Use only Zephyr specific POSIX header files so that the whole
system is getting values from the same files. There was an issue with
native_sim run of tests/net/socket/af_packet which had O_BLOCKING set
to 0x4000 from include/zephyr/posix/fcntl.h, but then the file
subsys/net/lib/socket/sockets.c was having O_BLOCKING set to 0x0800
because different header files were used.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Use always zephyr/posix/fcntl.h

Changing remaining users of fcntl.h to use the include from our own
POSIX file so that the values in there are consistent in all parts
of the sources.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

cfriedt commented:
@jukkar - #69979 was merged so this could use a rebase

cfriedt reviewed (dismissed):
Awesome - thank you @jukkar 🙌

rlubos reviewed (dismissed):
LGTM

cfriedt reviewed (dismissed):
Thank you @jukkar 🥳🙌

jukkar commented:
Thanks for the support :bowing_man:. There is still USB native_sim driver issues that needs fixing or a new driver, so at least one new version will be sent in hopefully near future.
@jfischer-no do you have time to look into the USB issue?

jukkar commented:
> should this PR go through another round of reviews, `s/Posix/POSIX/` would be nice (spotted in the migration guide blurb, maybe elsewhere?). Thanks!

Please add comments if you find something. I will collect them all and do an update at some point. I can do the Posix->POSIX already (but will not send a new version yet).

henrikbrixandersen reviewed (commented):
`samples/net/sockets/can/` changes OK.

kartben commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -170,6 +170,18 @@ Networking
   one IPv4 address / network interface, the netmask must be specified
   for each IPv4 address separately. (:github:`68419`)
 
+* Deprecate the :kconfig:option:`CONFIG_NET_SOCKETS_POSIX_NAMES` option. It is a legacy option
+  and was used to allow user to call BSD socket API while not enabling Posix API.
+  This could cause complications when building applications that wanted to enable the
+  :kconfig:option:`CONFIG_POSIX_API` option. This means that if the application wants to use
+  normal BSD socket interface, then it needs to enable :kconfig:option:`CONFIG_POSIX_API`.
+  If the application does not want or is not able to enable that option, then the socket API
+  calls need to be prefixed by a ``zsock_`` string.
+  All the sample applications that use BSD socket interface are changed to enable
+  :kconfig:option:`CONFIG_POSIX_API`. Internally the network stack will not enable Posix API
+  option which means that various network libraries that use sockets, are converted to
+  use the ``zsock_*`` API calls. (:github:`69950`)
+
```
s/Posix/POSIX/

jukkar commented:
* rebased against latest master
* resolving merge conflict
* updated code according to review comments

The native-posix/sim usb driver does not work and will fail the CI.

jukkar committed:
samples: net: Disable native_sim net USB driver in samples

The native_sim USB driver gives this error currently if one
enables CONFIG_POSIX_API.

drivers/usb/device/usb_dc_native_posix_adapt.c:22:10: \
 fatal error: sys/un.h: No such file or directory
   22 | #include <sys/un.h>
      |          ^~~~~~~~~~
compilation terminated.

Disable the netusb support from networking shells for native_sim
board for now so that CI can pass.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
Changes:
* fixed the `http_get` sample compilation
